File: Assets/Scripts/Networking/Message/ImageMessage.cs
using Networking.Message.Utils;
using UnityEngine;
using Utils.Extensions;

namespace Networking.Message
{
    /// <summary>
    /// Сообщение отправки изображения
    /// </summary>
    public class ImageMessage: IMessage
    {
        /// <summary>
        /// Тип сообщения. Требует 1 байт
        /// </summary>
        public MessageType MessageType { get; private set; }

        /// <summary>
        /// Идентификатор кадра, который мы отправили в MotionDetection. Требует 2 байта
        /// </summary>
        public ushort PacketId;

        /// <summary>
        /// Изображение.
        /// Передается размер (Требует на каждую координату по 2 байта)
        /// и сами байты изображения
        /// </summary>
        public Texture2D Image;

        public ImageMessage(MessageType messageType)
        {
            MessageType = messageType;
        }
        
        /// <summary>
        /// Сериализует класс в массив байтов 
        /// </summary>
        public byte[] Serialize()
        {
            var imageBytes = Image.EncodeToJPG();
            //MessageType + PacketId + (Image.Width + Image.Height + Image.JPG.Length)
            var length = (ushort) (1 + 2 + (2 + 2 + imageBytes.Length));
            this.CreateMessage(length, out var data);

            var offset = MessageExtensions.HEADER_LENGTH;
            offset = MessageExtensions.SetByte(in data, (byte) MessageType, offset);
            offset = MessageExtensions.SetBytes(in data, PacketId, offset);
            offset = MessageExtensions.SetBytes(in data, (ushort) Image.width, offset);
            offset = MessageExtensions.SetBytes(in data, (ushort) Image.height, offset);
            MessageExtensions.SetBytes(in data, imageBytes, offset);
            
            return data;
        }

        /// <summary>
        /// Десериализирует данные из массива байтов в класс 
        /// </summary>
        public static IMessage Deserialize(in byte[] data)
        {
            var offset = MessageExtensions.HEADER_LENGTH;
            var message = new ImageMessage((MessageType) data[offset++])
            {
                PacketId = MessageExtensions.GetUInt16(data, ref offset)
            };
            
            var width = MessageExtensions.GetUInt16(data, ref offset);
            var height = MessageExtensions.GetUInt16(data, ref offset);
            message.Image = new Texture2D(width, height, TextureFormat.RGBA32,false);
            message.Image.LoadImage(MessageExtensions.GetBytes(data, ref offset));
            
            return message;    
        }
    }
}

File: Python/Socket/Messages/ImageMessage.py
from Socket.Messages.Utils.MessageTypes import MessageTypes
import Socket.Utils.ByteConverter as ByteConverter
from Socket.Messages.Message import Message


# Класс сообщения картинки с одной из камер
class ImageMessage(Message):
    PacketId: int  # ushort : 2
    Width: int  # ushort : 2
    Height: int  # ushort : 2
    JpgImageData: bytearray

    def __init__(self, data: bytearray):
        self.MessageLength = ByteConverter.GetInteger(data)
        self.MessageType = MessageTypes(ByteConverter.GetInteger(data, 4, 1))
        self.PacketId = ByteConverter.GetInteger(data, 5, 2, False)
        self.Width = ByteConverter.GetInteger(data, 7, 2, False)
        self.Height = ByteConverter.GetInteger(data, 9, 2, False)
        # ToDo: Get image and convert it
        self.JpgImageData = ByteConverter.GetBytes(data, 11)

    def Serialize(self):
        returnArray = bytearray()
        length = 4 + 1 + 2 + (2 + 2 + len(self.JpgImageData))
        returnArray.extend(ByteConverter.GetBytesFromInteger(length))
        returnArray.extend(ByteConverter.GetBytesFromByteInteger(self.MessageType))
        returnArray.extend(ByteConverter.GetBytesFromUshort(self.PacketId))
        returnArray.extend(ByteConverter.GetBytesFromUshort(self.Width))
        returnArray.extend(ByteConverter.GetBytesFromUshort(self.Height))
        # ToDo: Get image and convert it
        returnArray.extend(self.JpgImageData)
        return returnArray

    @staticmethod
    def Deserialize(data: bytearray):
        return ImageMessage(data)